File: src/java/foundgine-runtime/src/main/java/com/foundgine/runtime/routing/RoutingContext.java
package com.foundgine.runtime.routing;
import com.foundgine.core.semantic.security.execution.SecurityExecutionContext;
import com.foundgine.runtime.controlplane.risk.RiskScore;
import java.util.*;
public record RoutingContext(String toolName,SecurityExecutionContext security,RiskScore riskScore,Map<String,Object> hints){public RoutingContext{hints=Map.copyOf(hints==null?Map.of():hints);}public Optional<Object> hint(String key){return Optional.ofNullable(hints.get(key));}}

File: src/csharp/Foundgine.Runtime/Routing/RoutingContext.cs
using Foundgine.Core.Semantic.Security.Execution;
using Foundgine.Runtime.ControlPlane.RiskScoring;

namespace Foundgine.Runtime.Routing;

/// <summary>
/// Inputs available to a routing decision. Routing decides <em>how</em> a
/// tool call runs (mode, runtime location, worker assignment, lifecycle),
/// never <em>which backend</em> executes it — backend selection remains a
/// host/DI concern resolved through <c>IExecutionProvider</c>.
/// </summary>
public sealed record RoutingContext(
    string ToolName,
    SecurityExecutionContext Security,
    RiskScore RiskScore,
    IReadOnlyDictionary<string, object?>? Hints = null)
{
    public IReadOnlyDictionary<string, object?> EffectiveHints =>
        Hints ?? EmptyHints;

    private static readonly IReadOnlyDictionary<string, object?> EmptyHints =
        new Dictionary<string, object?>(StringComparer.Ordinal);

    public bool TryGetHint(string key, out object? value) =>
        EffectiveHints.TryGetValue(key, out value);
}